### PR TITLE
preserve field order in Data.Obj

### PR DESCRIPTION
**Pull request by mossprescott**
This is just a simple switch to use ListMap in Data.Obj, as we do basically everywhere else in the code where we have maps of fields.

This fixes two planner tests which were marked with TODOs referring to this exact issue. Since it was so simple to make the switch, I'm not sure why we left those in and didn't just fix it.

I have not yet updated the regression tests. There are about 10 or so failures, some obviously for field order, but others also which I haven't looked at yet.

Any reason not to make this switch?

### CODE REVIEW

**Comment by jdegoes** [comment]
:+1:

**Comment by sellout** [comment]
Looks sensible to me.

**Comment by mossprescott** [comment]
Thanks. I'll take a look at the test later on and hopefully wrap this up.

File: it/src/main/resources/tests/constantAndGrouped.test [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "name": "constant and a grouped value",
+  "backends": { "mongodb_2_6": "pending", "mongodb_3_0": "pending", "mongodb_read_only": "pending", "mongodb_3_2": "pending" },
```
Comment (mossprescott) [inline it/src/main/resources/tests/constantAndGrouped.test]: This is a legitimate bad result that was not being caught by `containsExactly`. The result is 29k copies of the expected result, because we don't insert an `Arbitrary` reduction around a constant, but we probably should so I left the test as is and marked it pending.

I think we should go ahead and fix that, but perhaps not in this PR.

Comment (jdegoes) [inline it/src/main/resources/tests/constantAndGrouped.test]: It will be fixed by @sellout's PR, because `Constantly` will no longer be wrapped around literals, therefore, there is no need to insert an arbitrary reduction around them.

Comment (mossprescott) [inline it/src/main/resources/tests/constantAndGrouped.test]: Even better.

**Comment by mossprescott** [comment]
Rebased so I could fix the new test @sellout added.

File: it/src/main/resources/tests/constantAndGrouped.test (hunk 1 above)
Comment (sellout) [inline it/src/main/resources/tests/constantAndGrouped.test]: The PR mentioned above (#1142) has been merged, so if you rebase, you can see if that fixed it.

Comment (mossprescott) [inline it/src/main/resources/tests/constantAndGrouped.test]: Thanks, I'll rebase and unleash that test.

Comment (mossprescott) [inline it/src/main/resources/tests/constantAndGrouped.test]: Hmm. Doesn't seem to be any fixed-er.

**Comment by mossprescott** [comment]
@sellout: This is ready for a look, finally.

With the ControlService tests skipped, all four builds passed on the first try in Travis, and I just kicked it off again to make sure.

**Comment by mossprescott** [comment]
@sellout: Built clean in Travis with that fix in `objectConcat`. Thanks!

**Comment by sellout** [comment]
This looks great. Glad to finally fix all the issues plaguing integration tests. And happy to see some arbitraries becoming more arbitrary.

File: it/src/test/scala/quasar/ResultFileQueryRegressionSpec.scala [hunk 2]
```diff
@@ -45,7 +43,7 @@ class ResultFileQueryRegressionSpec
     for {
       tmpFile <- hoistM(manage.tempFile(DataDir)).liftM[Process]
       outFile <- query.executeQuery(expr, vars, tmpFile).liftM[Process]
-      cleanup =  hoistM(manage.delete(tmpFile))
+      cleanup =  hoistM(manage.delete(tmpFile)).whenM(outFile ≟ tmpFile)
```
Comment (wemrysi) [inline it/src/test/scala/quasar/ResultFileQueryRegressionSpec.scala]: Why is this necessary?

File: it/src/test/scala/quasar/regression/QueryRegressionTest.scala [hunk 3]
```diff
@@ -44,6 +44,9 @@ abstract class QueryRegressionTest[S[_]: Functor](
 
   import QueryRegressionTest._
 
+  // NB: forcing the examples to not overlap seems to avoid some errors in Travis
+  sequential
```
Comment (wemrysi) [inline it/src/test/scala/quasar/regression/QueryRegressionTest.scala]: Any idea why this is, or which examples? Is this something new required as a result of other changes, or had the failures been happening all along?

File: it/src/test/scala/quasar/ResultFileQueryRegressionSpec.scala (hunk 2 above)
Comment (sellout) [inline it/src/test/scala/quasar/ResultFileQueryRegressionSpec.scala]: Because sometimes the tmpFile never gets created.

Comment (wemrysi) [inline it/src/test/scala/quasar/ResultFileQueryRegressionSpec.scala]: Ah, right (for some reason I read that operator as "not equal to").

File: it/src/test/scala/quasar/regression/QueryRegressionTest.scala (hunk 3 above)
Comment (mossprescott) [inline it/src/test/scala/quasar/regression/QueryRegressionTest.scala]: It's a bit mysterious, but when I changed the matchers to be more rigorous about verifying the results and capturing failures, I started seeing "137" errors in Travis. According to [this issue](https://github.com/travis-ci/travis-ci/issues/4192#issuecomment-119053081) @jr dug up, it may be related to using more memory. My best guess is that the tests, when running in parallel, are now overlapping more than they used to, and that's just too much for the under-powered Travis Docker instances in some way.

Another hypothesis would be that the matchers are no longer "memory safe" in that they somehow keep an unbounded number of values from the query result in memory until they're done reading. I convinced myself that wasn't the case, but another look wouldn't hurt.

In any case, I tried it several times and `sequential` _always_ made the error go away, and does not noticably slow them down.

**Comment by sellout** [comment]
This is in ready – which I should have set it to last week, but just did this morning.

**Comment by mossprescott** [comment]
Thanks!

I will revert the change to skip the ControlService tests, since I haven't noticed as many of those failures lately.

**Comment by mossprescott** [comment]
Merged as 5.0.7.
